test/e2e/node/container_readiness_during_termination.go

```diff
@@ -0,0 +1,259 @@
+		})
+		framework.ExpectNoError(err, "pod should be fully terminated")
+	})
+
+	ginkgo.It("should update readiness when container crashes during termination", func(ctx context.Context) {
+		ginkgo.By("Creating a pod with container that will crash")
+		podName := "test-pod-liveness-readiness-" + string(uuid.NewUUID())
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      podName,
+				Namespace: ns,
+			},
+			Spec: v1.PodSpec{
+				TerminationGracePeriodSeconds: func() *int64 { v := int64(120); return &v }(),
+				Containers: []v1.Container{
+					{
+						Name:  "test-container",
+						Image: imageutils.GetE2EImage(imageutils.Agnhost),
+						Args:  []string{"sh", "-c", "sleep 10; echo 'Container crashing...'; exit 1"},
+						Lifecycle: &v1.Lifecycle{
+							PreStop: &v1.LifecycleHandler{
+								Exec: &v1.ExecAction{
+									Command: []string{"sh", "-c", "sleep 30"},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		ginkgo.By("Creating the pod")
+		pod = podClient.Create(ctx, pod)
+
+		ginkgo.By("Waiting for pod to be running")
+		err := e2epod.WaitForPodRunningInNamespace(ctx, cs, pod)
+		framework.ExpectNoError(err, "pod should be running")
+
+		ginkgo.By("Waiting for container to crash and restart")
+		var containerRestarted bool
+		err = wait.PollUntilContextTimeout(ctx, 2*time.Second, 1*time.Minute, true, func(ctx context.Context) (bool, error) {
+			pod, err := podClient.Get(ctx, pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+
+			status := pod.Status.ContainerStatuses[0]
+			if status.RestartCount > 0 {
+				containerRestarted = true
+				framework.Logf("Container restarted, restart count: %d", status.RestartCount)
+				return true, nil
+			}
+
+			return false, nil
+		})
+		framework.ExpectNoError(err, "container should restart after crash")
+		gomega.Expect(containerRestarted).To(gomega.BeTrueBecause("container should have restarted"))
+
+		ginkgo.By("Deleting the pod during container restart")
+		err = podClient.Delete(ctx, pod.Name, metav1.DeleteOptions{})
+		framework.ExpectNoError(err, "failed to delete pod")
+
+		ginkgo.By("Monitoring readiness during termination with preStop hook")
+		var readinessUpdated bool
+		err = wait.PollUntilContextTimeout(ctx, 2*time.Second, 2*time.Minute, true, func(ctx context.Context) (bool, error) {
+			pod, err := podClient.Get(ctx, pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+
+			if pod.DeletionTimestamp == nil {
+				return false, fmt.Errorf("pod should be terminating")
+			}
+
+			readyContainers := 0
+			for _, status := range pod.Status.ContainerStatuses {
+				if status.Ready {
+					readyContainers++
+				}
+			}
+
+			framework.Logf("Pod status during termination: %d/%d containers ready", readyContainers, len(pod.Status.ContainerStatuses))
+
+			// If container is being killed due to crash, readiness should be updated
+			if readyContainers < len(pod.Status.ContainerStatuses) {
```

Comment:
so this test will succeed if both containers will be marked as NotReady right away? Should we expect 1 container being not ready?